src/validator/MatchValidator.cpp

```diff
@@ -694,44 +695,68 @@ Status MatchValidator::validateGroup(YieldClauseContext &yieldCtx) const {
     for (auto *col : cols) {
         auto *colExpr = col->expr();
         auto colOldName = col->name();
-        if (colExpr->kind() != Expression::Kind::kAggregate) {
-            auto collectAggCol = colExpr->clone();
-            auto aggs =
-                ExpressionUtils::collectAll(collectAggCol, {Expression::Kind::kAggregate});
-            for (auto *agg : aggs) {
-                DCHECK_EQ(agg->kind(), Expression::Kind::kAggregate);
-                if (!ExpressionUtils::checkAggExpr(static_cast<const AggregateExpression *>(agg))
-                         .ok()) {
-                    return Status::SemanticError("Aggregate function nesting is not allowed: `%s'",
-                                                 colExpr->toString().c_str());
+        auto aggs = ExpressionUtils::collectAll(colExpr, {Expression::Kind::kAggregate});
+        for (auto *agg : aggs) {
+            DCHECK_EQ(agg->kind(), Expression::Kind::kAggregate);
+            if (!ExpressionUtils::checkAggExpr(static_cast<const AggregateExpression *>(agg))
+                     .ok()) {
+                return Status::SemanticError("Aggregate function nesting is not allowed: `%s'",
+                                             colExpr->toString().c_str());
+            }
+        }
+    }
+
+    yieldCtx.needGenProject_ = true;
+    for (auto *col : cols) {
+        auto *colExpr = col->expr();
+        auto colOldName = col->toString();
+        auto foldRes = ExpressionUtils::foldConstantExpr(colExpr);
+        NG_RETURN_IF_ERROR(foldRes);
+        auto *newExpr = foldRes.value();
+
+        if (newExpr->kind() == Expression::Kind::kAggregate) {
+            yieldCtx.groupItems_.emplace_back(newExpr);
+            yieldCtx.aggOutputColumnNames_.emplace_back(newExpr->toString());
+            yieldCtx.projCols_->addColumn(new YieldColumn(
+                VariablePropertyExpression::make(pool, "", colOldName), colOldName));
+            yieldCtx.projOutputColumnNames_.emplace_back(colOldName);
+        } else {
+            AggregateVisitor visitor;
```

Comment:
Merge this if-else block. You always set these variables in the if-block, but don't set `groupKeys_`, isn’t it?

Reply:
```cpp
Status MatchValidator::validateGroup(YieldClauseContext &yieldCtx) const {
    NG_RETURN_IF_ERROR(YieldClauseContext::checkAggregation()); 
    YieldClauseContext::ExtractGroupItems(); // hiding the visitor's logic
   
    return Status::OK();
}

```